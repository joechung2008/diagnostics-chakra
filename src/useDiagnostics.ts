import { useMemo } from "react";

interface CacheEntry {
  data: Diagnostics | undefined;
  error: Error | undefined;
  promise: Promise<Diagnostics> | undefined;
}

const cache = new Map<string, CacheEntry>();

async function fetchDiagnosticsData(environment: string): Promise<Diagnostics> {
  const response = await fetch(environment);
  if (!response.ok) {
    throw new Error(`Failed to fetch diagnostics: ${response.statusText}`);
  }

  return response.json();
}

export function clearCache(): void {
  cache.clear();
}

export function useDiagnostics(environment: string): Diagnostics {
  const cacheKey = environment;

  const entry = useMemo(() => {
    if (!cache.has(cacheKey)) {
      cache.set(cacheKey, {
        promise: undefined,
        data: undefined,
        error: undefined,
      });
    }

    return cache.get(cacheKey)!;
  }, [cacheKey]);

  if (entry.error) {
    throw entry.error;
  }

  if (!entry.data) {
    if (!entry.promise) {
      entry.promise = fetchDiagnosticsData(environment);
      entry.promise
        .then((data) => {
          entry.data = data;
          entry.promise = undefined;
        })
        .catch((error) => {
          entry.error = error;
          entry.promise = undefined;
        });
    }

    throw entry.promise;
  }

  return entry.data;
}
