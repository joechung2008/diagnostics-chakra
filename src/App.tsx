import BuildInfo from "@/BuildInfo";
import { ColorModeButton } from "@/components/ui/ColorMode";
import Extension from "@/Extension";
import Extensions from "@/Extensions";
import ServerInfo from "@/ServerInfo";
import { clearCache, useDiagnostics } from "@/useDiagnostics";
import { isExtensionInfo } from "@/utils";
import { Box, Button, Flex, Menu, Portal, Tabs } from "@chakra-ui/react";
import { startTransition, useCallback, useMemo, useState } from "react";

type Environment = (typeof Environment)[keyof typeof Environment];

const Environment = {
  Public: "https://hosting.portal.azure.net/api/diagnostics",
  Fairfax: "https://hosting.azureportal.usgovcloudapi.net/api/diagnostics",
  Mooncake: "https://hosting.azureportal.chinacloudapi.cn/api/diagnostics",
} as const;

function getEnvironnmentName(environment: Environment | undefined): string {
  switch (environment) {
    case Environment.Public:
      return "Public Cloud";
    case Environment.Fairfax:
      return "Fairfax";
    case Environment.Mooncake:
      return "Mooncake";
    default:
      return "Select environment";
  }
}

const App: React.FC = () => {
  const [environments, setEnvironments] = useState<
    Record<string, Environment[]>
  >({
    environment: [Environment.Public],
  });
  const environment = useMemo<Environment>(
    () => environments.environment[0],
    [environments.environment]
  );
  const diagnostics = useDiagnostics(environment);
  const [extension, setExtension] = useState<ExtensionInfo>();
  const [selectedTab, setSelectedTab] = useState<string>("extensions");

  const handleLinkClick: ExtensionsProps["onLinkClick"] = useCallback(
    (_, item) => {
      if (item) {
        const $extension = diagnostics?.extensions[item.key];
        if (isExtensionInfo($extension)) {
          setExtension($extension);
        }
      }
    },
    [diagnostics?.extensions]
  );

  const showPaasServerless = useMemo(
    () => isExtensionInfo(diagnostics?.extensions["paasserverless"]),
    [diagnostics?.extensions]
  );

  const environmentName = useMemo(
    () => getEnvironnmentName(environment),
    [environment]
  );

  const handleEnvironmentChange = useCallback((value: string) => {
    if (value) {
      startTransition(() => {
        setEnvironments((previous) => ({
          ...previous,
          environment: [value as Environment],
        }));
        setExtension(undefined);
      });
      clearCache();
    }
  }, []);

  if (!diagnostics) {
    return null;
  }

  return (
    <Flex flexDirection="column" h="100vh" p="1">
      <Flex justify="space-between">
        <Flex gap="1">
          <Menu.Root>
            <Menu.Trigger asChild>
              <Button variant="outline">{environmentName}</Button>
            </Menu.Trigger>
            <Portal>
              <Menu.Positioner>
                <Menu.Content>
                  {Object.entries(Environment).map(([key, value]) => (
                    <Menu.Item
                      key={key}
                      value={value}
                      onClick={() => handleEnvironmentChange(value)}
                    >
                      {getEnvironnmentName(value)}
                    </Menu.Item>
                  ))}
                </Menu.Content>
              </Menu.Positioner>
            </Portal>
          </Menu.Root>
          {showPaasServerless && (
            <Button
              key="paasserverless"
              variant="ghost"
              onClick={() => {
                const paasserverless =
                  diagnostics?.extensions["paasserverless"];
                if (isExtensionInfo(paasserverless)) {
                  setExtension(paasserverless);
                }
              }}
            >
              paasserverless
            </Button>
          )}
          <Button
            key="websites"
            variant="ghost"
            onClick={() => {
              const websites = diagnostics?.extensions["websites"];
              if (isExtensionInfo(websites)) {
                setExtension(websites);
              }
            }}
          >
            websites
          </Button>
        </Flex>
        <ColorModeButton />
      </Flex>
      <Tabs.Root
        value={selectedTab}
        onValueChange={(details) => setSelectedTab(details.value)}
      >
        <Tabs.List>
          <Tabs.Trigger aria-controls="extensions-tab" value="extensions">
            Extensions
          </Tabs.Trigger>
          <Tabs.Trigger aria-controls="build-tab" value="build">
            Build Information
          </Tabs.Trigger>
          <Tabs.Trigger aria-controls="server-tab" value="server">
            Server Information
          </Tabs.Trigger>
        </Tabs.List>
      </Tabs.Root>
      {selectedTab === "extensions" && diagnostics?.extensions && (
        <Box id="extensions-tab" flex="1" overflowY="auto" role="tabpanel">
          <Flex flexDirection="row" gap="4" h="100%">
            <Extensions
              extensions={diagnostics.extensions}
              onLinkClick={handleLinkClick}
            />
            {extension && <Extension {...extension} />}
          </Flex>
        </Box>
      )}
      {selectedTab === "build" && diagnostics?.buildInfo && (
        <Box id="build-tab" flex="1" overflowY="auto" role="tabpanel">
          <BuildInfo {...diagnostics.buildInfo} />
        </Box>
      )}
      {selectedTab === "server" && diagnostics?.serverInfo && (
        <Box id="server-tab" flex="1" overflowY="auto" role="tabpanel">
          <ServerInfo {...diagnostics.serverInfo} />
        </Box>
      )}
    </Flex>
  );
};

export default App;
