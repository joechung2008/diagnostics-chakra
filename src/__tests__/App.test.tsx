import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import { beforeEach, describe, expect, it, vi } from "vitest";
import App from "../App";
import { useDiagnostics } from "../useDiagnostics";
import { TestWrapper } from "./test-utils";

// Mock the useDiagnostics hook
vi.mock("../useDiagnostics", () => ({
  useDiagnostics: vi.fn(),
  clearCache: vi.fn(),
}));

const mockUseDiagnostics = vi.mocked(useDiagnostics);

const mockDiagnosticsData = {
  buildInfo: { buildVersion: "1.2.3" },
  extensions: {
    testExtension: {
      extensionName: "Test Extension",
      config: { key1: "value1" },
      stageDefinition: { stage1: ["step1"] },
    },
    paasserverless: {
      extensionName: "PaaS Serverless",
      config: { enabled: "true" },
    },
    websites: {
      extensionName: "Websites",
      config: { version: "2.0" },
    },
  },
  serverInfo: {
    deploymentId: "deploy-123",
    extensionSync: { totalSyncAllCount: 42 },
    hostname: "server.example.com",
    nodeVersions: "v18.17.0",
    serverId: "server-456",
    uptime: 1234567890,
  },
};

describe("App", () => {
  beforeEach(() => {
    vi.clearAllMocks();
    mockUseDiagnostics.mockReturnValue(mockDiagnosticsData);
  });

  it("should render loading state when diagnostics is null", () => {
    mockUseDiagnostics.mockReturnValue(null as never);

    render(
      <TestWrapper>
        <App />
      </TestWrapper>
    );

    // When diagnostics is null, App returns null, so main app elements should not be present
    expect(
      screen.queryByRole("button", { name: "Public Cloud" })
    ).not.toBeInTheDocument();
    expect(
      screen.queryByRole("tab", { name: "Extensions" })
    ).not.toBeInTheDocument();
  });

  it("should render app with default environment", () => {
    const { asFragment } = render(
      <TestWrapper>
        <App />
      </TestWrapper>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should display environment selector with default value", () => {
    act(() => {
      render(
        <TestWrapper>
          <App />
        </TestWrapper>
      );
    });

    expect(
      screen.getByRole("button", { name: "Public Cloud" })
    ).toBeInTheDocument();
  });

  it("should render tabs for navigation", () => {
    act(() => {
      render(
        <TestWrapper>
          <App />
        </TestWrapper>
      );
    });

    expect(screen.getByRole("tab", { name: "Extensions" })).toBeInTheDocument();
    expect(
      screen.getByRole("tab", { name: "Build Information" })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("tab", { name: "Server Information" })
    ).toBeInTheDocument();
  });

  it("should display extensions tab by default", () => {
    act(() => {
      render(
        <TestWrapper>
          <App />
        </TestWrapper>
      );
    });

    // Should show extensions content by default
    expect(screen.getByText("Test Extension")).toBeInTheDocument();
  });

  it("should switch to build information tab", async () => {
    render(
      <TestWrapper>
        <App />
      </TestWrapper>
    );

    await act(async () => {
      fireEvent.click(screen.getByRole("tab", { name: "Build Information" }));
    });

    await waitFor(() => {
      expect(screen.getByText("Build Version")).toBeInTheDocument();
    });
    expect(screen.getByText("1.2.3")).toBeInTheDocument();
  });

  it("should switch to server information tab", async () => {
    render(
      <TestWrapper>
        <App />
      </TestWrapper>
    );

    await act(async () => {
      fireEvent.click(screen.getByRole("tab", { name: "Server Information" }));
    });

    await waitFor(() => {
      expect(screen.getByText("Hostname")).toBeInTheDocument();
    });
    expect(screen.getByText("server.example.com")).toBeInTheDocument();
  });

  it("should show paasserverless button when extension exists", () => {
    act(() => {
      render(
        <TestWrapper>
          <App />
        </TestWrapper>
      );
    });

    expect(
      screen.getByRole("button", { name: "paasserverless" })
    ).toBeInTheDocument();
  });

  it("should show websites button", () => {
    act(() => {
      render(
        <TestWrapper>
          <App />
        </TestWrapper>
      );
    });

    expect(
      screen.getByRole("button", { name: "websites" })
    ).toBeInTheDocument();
  });

  it("should handle extension selection from extensions list", async () => {
    render(
      <TestWrapper>
        <App />
      </TestWrapper>
    );

    // Click on an extension in the extensions list
    await act(async () => {
      fireEvent.click(screen.getByText("Test Extension"));
    });

    // Should display the extension details
    await waitFor(() => {
      expect(screen.getByText("Test Extension")).toBeInTheDocument();
    });
  });

  it("should handle paasserverless button click", () => {
    render(
      <TestWrapper>
        <App />
      </TestWrapper>
    );

    act(() => {
      fireEvent.click(screen.getByRole("button", { name: "paasserverless" }));
    });

    // Should display paasserverless extension details - look for the heading
    expect(
      screen.getByRole("heading", { name: "PaaS Serverless" })
    ).toBeInTheDocument();
  });

  it("should handle websites button click", () => {
    render(
      <TestWrapper>
        <App />
      </TestWrapper>
    );

    act(() => {
      fireEvent.click(screen.getByRole("button", { name: "websites" }));
    });

    // Should display websites extension details - look for the heading
    expect(
      screen.getByRole("heading", { name: "Websites" })
    ).toBeInTheDocument();
  });

  it("should handle environment change", () => {
    render(
      <TestWrapper>
        <App />
      </TestWrapper>
    );

    // Open environment menu
    act(() => {
      fireEvent.click(screen.getByRole("button", { name: "Public Cloud" }));
    });

    // Since menu items are in a Portal, we need to wait for them to appear
    // or test the menu trigger functionality
    expect(
      screen.getByRole("button", { name: "Public Cloud" })
    ).toBeInTheDocument();

    // The clearCache should be called when environment changes
    // This test verifies the menu can be opened
  });

  it("should handle environment change to Mooncake", () => {
    render(
      <TestWrapper>
        <App />
      </TestWrapper>
    );

    // Open environment menu
    act(() => {
      fireEvent.click(screen.getByRole("button", { name: "Public Cloud" }));
    });

    // Test that menu trigger is working
    expect(
      screen.getByRole("button", { name: "Public Cloud" })
    ).toBeInTheDocument();
  });

  it("should render color mode button", () => {
    act(() => {
      render(
        <TestWrapper>
          <App />
        </TestWrapper>
      );
    });

    // Color mode button should be present
    expect(
      screen.getByRole("button", { name: /color mode/i })
    ).toBeInTheDocument();
  });

  it("should handle tab switching correctly", async () => {
    render(
      <TestWrapper>
        <App />
      </TestWrapper>
    );

    // Start on extensions tab
    expect(screen.getByText("Test Extension")).toBeInTheDocument();

    // Switch to build tab
    await act(async () => {
      fireEvent.click(screen.getByRole("tab", { name: "Build Information" }));
    });
    await waitFor(() => {
      expect(screen.getByText("Build Version")).toBeInTheDocument();
    });

    // Switch to server tab
    await act(async () => {
      fireEvent.click(screen.getByRole("tab", { name: "Server Information" }));
    });
    await waitFor(() => {
      expect(screen.getByText("Hostname")).toBeInTheDocument();
    });

    // Switch back to extensions
    await act(async () => {
      fireEvent.click(screen.getByRole("tab", { name: "Extensions" }));
    });
    await waitFor(() => {
      expect(screen.getByText("Test Extension")).toBeInTheDocument();
    });
  });
});
