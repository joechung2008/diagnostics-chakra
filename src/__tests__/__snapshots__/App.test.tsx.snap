// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > should render app with default environment 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="css-1l1i41u"
  >
    <div
      class="css-gg4vpm"
    >
      <div
        class="css-11ae5ff"
      >
        <button
          aria-haspopup="menu"
          class="chakra-button chakra-menu__trigger css-1ocnt5x"
          data-part="trigger"
          data-scope="menu"
          data-state="closed"
          data-uid="«r0»"
          dir="ltr"
          id="menu:«r0»:trigger"
          type="button"
        >
          Public Cloud
        </button>
        <button
          class="chakra-button css-1mi07d4"
          type="button"
        >
          paasserverless
        </button>
        <button
          class="chakra-button css-1mi07d4"
          type="button"
        >
          websites
        </button>
      </div>
      <button
        aria-label="Toggle color mode"
        class="chakra-button css-frz056"
        type="button"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="4"
          />
          <path
            d="M12 2v2"
          />
          <path
            d="M12 20v2"
          />
          <path
            d="m4.93 4.93 1.41 1.41"
          />
          <path
            d="m17.66 17.66 1.41 1.41"
          />
          <path
            d="M2 12h2"
          />
          <path
            d="M20 12h2"
          />
          <path
            d="m6.34 17.66-1.41 1.41"
          />
          <path
            d="m19.07 4.93-1.41 1.41"
          />
        </svg>
      </button>
    </div>
    <div
      class="chakra-tabs__root css-upaspd"
      data-orientation="horizontal"
      data-part="root"
      data-scope="tabs"
      dir="ltr"
      id="tabs:«r1»"
    >
      <div
        aria-orientation="horizontal"
        class="chakra-tabs__list css-yamsmp"
        data-orientation="horizontal"
        data-part="list"
        data-scope="tabs"
        dir="ltr"
        id="tabs:«r1»:list"
        role="tablist"
      >
        <button
          aria-controls="tabs:«r1»:content-extensions"
          aria-selected="true"
          class="chakra-tabs__trigger css-19t98pl"
          data-focus=""
          data-orientation="horizontal"
          data-ownedby="tabs:«r1»:list"
          data-part="trigger"
          data-scope="tabs"
          data-selected=""
          data-ssr=""
          data-value="extensions"
          dir="ltr"
          id="tabs:«r1»:trigger-extensions"
          role="tab"
          tabindex="0"
          type="button"
        >
          Extensions
        </button>
        <button
          aria-selected="false"
          class="chakra-tabs__trigger css-19t98pl"
          data-orientation="horizontal"
          data-ownedby="tabs:«r1»:list"
          data-part="trigger"
          data-scope="tabs"
          data-ssr=""
          data-value="build"
          dir="ltr"
          id="tabs:«r1»:trigger-build"
          role="tab"
          tabindex="-1"
          type="button"
        >
          Build Information
        </button>
        <button
          aria-selected="false"
          class="chakra-tabs__trigger css-19t98pl"
          data-orientation="horizontal"
          data-ownedby="tabs:«r1»:list"
          data-part="trigger"
          data-scope="tabs"
          data-ssr=""
          data-value="server"
          dir="ltr"
          id="tabs:«r1»:trigger-server"
          role="tab"
          tabindex="-1"
          type="button"
        >
          Server Information
        </button>
      </div>
    </div>
    <div
      class="css-177dxab"
    >
      <div
        class="css-e3v19i"
      >
        <nav
          aria-label="Extensions"
          class="css-rqcg1t"
        >
          <button
            class="chakra-button css-1tlpiy1"
            type="button"
          >
            PaaS Serverless
          </button>
          <button
            class="chakra-button css-1tlpiy1"
            type="button"
          >
            Test Extension
          </button>
          <button
            class="chakra-button css-1tlpiy1"
            type="button"
          >
            Websites
          </button>
        </nav>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
